_posts/2023-08-28-practical-type-level-programming-part2.md

```diff
@@ -0,0 +1,496 @@
+---
+title: "Practical Type-Level Programming in Scala 3, Part 2: Serializing a Concrete Class"
+date: 2023-08-28
+header:
+  image: "/images/blog cover.jpg"
+tags: []
+excerpt: "JSON serialization of a concrete class with type-level techniques."
+---
+
+Welcome to the second installation of this series. I'm assuming that you've read the [first part](/practical-type-level-programming-part1), and are familiar with the "flat JSON" problem we are trying to solve.
+
+Let us now proceed to flat serializing a concrete class. This will allow us to dip our toes into a bit of type-level programming without going into the deep end straight away.
+
+## Serializing a Concrete Class
+
+Since we'll be dealing with JSON, we'll need a JSON library to do the JSON work for us. I'll be using [Circe](https://github.com/circe/circe), but pretty much any other Scala JSON library (of which there are plenty) will do just fine. I won't spend too much time explaining the mechanics of Circe, in the hope that the JSON code we'll be writing is sufficiently self-explanatory.
```

Comment:
As mentioned in another comment, I'm using scala-cli and providing a link to the configuration file ([here](https://github.com/rockthejvm/rockthejvm.github.io/pull/39/files/acb6ab145443e5308b53021d3c8a770850a61761#diff-2c16bd4a30182a36770712dbe3d9e98c365fec0d8c07b5da86e260a959c4a39fR18)).

Will that be good enough? Or do you have a strong preference for using sbt?

Reply:
Any tool is okay, I'm good with Scala CLI or SBT or other things, as long as we specify it at the beginning of the article so that the reader can follow along in the same tool

Reply:
Will make sure the instructions are there.